cmd/zoekt-sourcegraph-indexserver/merge.go

```diff
@@ -30,62 +30,90 @@ var metricShardMergingDuration = promauto.NewHistogramVec(prometheus.HistogramOp
 	Buckets: prometheus.LinearBuckets(30, 30, 10),
 }, []string{"error"})
 
-// doMerge drives the merge process.
-func doMerge(dir string, targetSizeBytes int64, simulate bool) error {
+func pickCandidates(shards []candidate, targetSizeBytes int64) compound {
+	c := compound{}
+	for _, shard := range shards {
+		c.add(shard)
+		if c.size >= targetSizeBytes {
+			return c
+		}
+	}
+	return compound{}
+}
+
+var mergeRunning atomic.Bool
+
+func defaultMergeCmd(args ...string) *exec.Cmd {
+	cmd := exec.Command("zoekt-merge-index", "merge")
+	cmd.Args = append(cmd.Args, args...)
+	return cmd
+}
+
+// doMerge drives the merge process. It holds the lock on s.indexDir for the
+// duration of 1 merge, which might be several minutes, depending on the target
+// size of the compound shard.
+func (s *Server) doMerge() {
+	s.merge(defaultMergeCmd)
+}
+
+// same as doMerge but with a configurable merge command.
+func (s *Server) merge(mergeCmd func(args ...string) *exec.Cmd) {
+
+	// Guard against the user triggering competing merge jobs with the debug
+	// command.
+	if !mergeRunning.CAS(false, true) {
+		debug.Printf("merge already running\n")
+		return
+	}
+	defer mergeRunning.Store(false)
+
 	metricShardMergingRunning.Set(1)
 	defer metricShardMergingRunning.Set(0)
 
 	wc := &lumberjack.Logger{
-		Filename:   filepath.Join(dir, "zoekt-merge-log.tsv"),
+		Filename:   filepath.Join(s.IndexDir, "zoekt-merge-log.tsv"),
 		MaxSize:    100, // Megabyte
 		MaxBackups: 5,
 	}
 
-	if simulate {
-		debug.Println("simulating")
-	}
+	// We keep creating compound shards until we run out of shards to merge or until
+	// we encounter an error during merging.
+	next := true
+	for next {
+		s.muIndexDir.Global(func() {
+			candidates, excluded := loadCandidates(s.IndexDir)
+			debug.Printf("loadCandidates: candidates=%d excluded=%d\n", len(candidates), excluded)
 
-	shards, excluded := loadCandidates(dir)
-	debug.Printf("merging: found %d candidate shards, %d shards were excluded\n", len(shards), excluded)
-	if len(shards) == 0 {
-		return nil
-	}
+			c := pickCandidates(candidates, s.TargetSizeBytes)
+			if len(c.shards) <= 1 {
+				debug.Printf("could not find enough shards to build a compound shard\n")
+				next = false
+				return
+			}
+			debug.Printf("start merging: shards=%d total_size=%.2fMiB\n", len(c.shards), float64(c.size)/(1024*1024))
```

Comment:
I would make all the logs here a higher log level. merging is slow and runs irregularly, so I believe it won't be spammy to have this info in the logs.